### PR TITLE
Refactor gradient accumulation (GA) and vocab tiling (VT)

### PR DESCRIPTION
**Pull request by NuojCheng**
# Description

Refactor some tiling-related functions in maxtext. This PR includes following things:

- Move gradient accumulation (GA) codes to gradient_accumulation.py
- Update the reshape function in GA. The previous one may lead to additional all-to-all.
- Move vocab tiling (VT) codes to vocabulary_tiling.py
- Add one additional test for GA
- Make some VT tests scheduled only

# Tests

Standard MaxText tests + one additional GA test in tiling_test.py.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/maxtext_utils.py [hunk 1]
```diff
@@ -97,14 +99,14 @@ def get_shaped_batch(config):
 
 def should_prevent_cse_in_remat(config):
   """Determines whether to prevent common subexpression elimination (CSE) in remat.
-  
+
```
Comment (richjames0) [inline src/MaxText/maxtext_utils.py:102]: nit: thanks for fixing whitespace but could this possibly be done in another PR? some of us use blame a lot so this can be helpful

File: src/MaxText/maxtext_utils.py [hunk 2]
```diff
@@ -1000,10 +1004,12 @@ def add_data_to_sharding(mesh, path, aval, sharding):
   try:
     sharded_shape = sharding.shard_shape(aval.shape)
   except Exception as e:
-    raise AssertionError(f"Could not shard value {jax.tree_util.keystr(path)} of shape={aval.shape} with {sharding=}") from e
+    raise AssertionError(
```
Comment (richjames0) [inline src/MaxText/maxtext_utils.py]: same as the whitespace comment above and also on the single vs. double quote changes below

File: src/MaxText/maxtext_utils.py [hunk 3]
```diff
@@ -1042,17 +1049,22 @@ def maybe_update_params_sharding_with_opt(config, state_mesh_shardings):
...
     if isinstance(state_mesh_shardings.opt_state, optax.ScaleByAdamState):
       sharded_fp32_params = state_mesh_shardings.opt_state.mu
-    elif isinstance(state_mesh_shardings.opt_state, tuple) and isinstance(state_mesh_shardings.opt_state[0], optax.ScaleByAdamState):
+    elif isinstance(state_mesh_shardings.opt_state, tuple) and isinstance(
+        state_mesh_shardings.opt_state[0], optax.ScaleByAdamState
+    ):
       sharded_fp32_params = state_mesh_shardings.opt_state[0].mu
     else:
-      raise NotImplementedError(f"Could not find optimizer state shardings from optimizer of type {type(state_mesh_shardings.opt_state)}")
+      raise NotImplementedError(
```
Comment (richjames0) [inline src/MaxText/maxtext_utils.py]: i find it strictly worse to split these long lines over three lines but acknowledge we do need to have some limit and apparently some of these lines broke it (i think it's 120 chars iirc). could you look at getting these back to one liners e.g. by factoring out intermediate variables (granted that adds a line but 2 < 3) or making the text more terse

File: src/MaxText/maxtext_utils.py [hunk 4]
```diff
@@ -1145,7 +1157,11 @@ def get_abstract_state(model, tx, config, rng, mesh, is_training=True):
   if is_training and config.shard_optimizer_over_data:
     # Add data to sharding for optimizer state
     state_mesh_shardings = state_mesh_shardings.replace(
-      opt_state=jax.tree.map_with_path(functools.partial(add_data_to_sharding, mesh), unbox_logicallypartioned(abstract_state).opt_state, state_mesh_shardings.opt_state)
+        opt_state=jax.tree.map_with_path(
```
Comment (richjames0) [inline src/MaxText/maxtext_utils.py:1156]: indentation has changed here i think?

File: src/MaxText/maxtext_utils.py [hunk 5]
```diff
@@ -1542,253 +1558,3 @@ def all_gather_over_fsdp(variables, sharding_info, mesh, logical_axis_rules):
...
-    config: The model and training configuration.
-    model: The NNX model instance.
-    params: The model parameters.
-    is_train: A boolean indicating if the model is in training mode.
-
-  Returns:
-    The total cross-entropy loss.
-  """
-  one_hot_targets = jax.nn.one_hot(data["targets"], config.vocab_size)
-  xent, _ = max_utils.cross_entropy_with_logits(logits, one_hot_targets, 0.0)
```
Comment (richjames0) [inline src/MaxText/maxtext_utils.py:1580]: nit: i'm not a fan of naming every parameter in the call but it's not clear here what 0.0 is so maybe you could put a name to that one, e.g. `one_hot_targets, magic_constant=0.0`

Comment (richjames0) [reply src/MaxText/maxtext_utils.py:1580]: same applies in linen fn btw

Comment (richjames0) [reply src/MaxText/maxtext_utils.py:1580]: oh haha these were moved per my request - thanks! but same will probably apply in the new file

**Review by richjames0 — COMMENTED** [review]
nit: could you s/GA/gradient accumulation in the title? it will appear outside the context of the body in many places and many won't make that leap

thank you!

File: src/MaxText/maxtext_utils.py (hunk 1 above)
Comment (NuojCheng) [reply src/MaxText/maxtext_utils.py:102]: Already fixed in #2511 , will rebase

File: src/MaxText/maxtext_utils.py (hunk 4 above)
Comment (NuojCheng) [reply src/MaxText/maxtext_utils.py:1156]: TY. Will fix

File: src/MaxText/maxtext_utils.py (hunk 2 above)
Comment (NuojCheng) [reply src/MaxText/maxtext_utils.py]: I will rebase with #2511

File: src/MaxText/maxtext_utils.py (hunk 3 above)
Comment (NuojCheng) [reply src/MaxText/maxtext_utils.py]: will do

File: src/MaxText/maxtext_utils.py (hunk 5 above)
Comment (NuojCheng) [reply src/MaxText/maxtext_utils.py:1580]: will do

File: src/MaxText/tiling_utils.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+    is_train: A boolean indicating if the model is in training mode.
+
+  Returns:
+    The total cross-entropy loss.
+  """
+  if config.num_vocab_tiling > 1:
+    hidden_state_key = ("intermediates", "decoder", "hidden_states")
+    hidden_states = maxtext_utils.get_nested_value(intermediate_outputs, hidden_state_key)[0]
+    total_loss = get_vocab_tiling_loss_linen(hidden_states, data, config, model, params, is_train)
+  else:
```
Comment (gobbleturk) [inline src/MaxText/tiling_utils.py]: This is the main loss function of train.py (in the case of num_Vocab_tiling=1). There is no tiling in this "else" branch. It may be surprising/confusing/poor organization to have the main non-tile loss function in a file called tiling_utils.py

**Review by gobbleturk — APPROVED** [review]

File: src/MaxText/tiling_utils.py [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+
+from flax import linen as nn
+
+import jax
+import jax.numpy as jnp
+from MaxText import max_utils
+from MaxText import maxtext_utils
+
+
+# Vocab Tiling Helper Functions
```
Comment (gobbleturk) [inline src/MaxText/tiling_utils.py]: This comment and the method name are a severe mismatch - I would recommend some organization like

compute_loss_nnx defined in some more generic file (maxtext_utils or a new loss_fn file or something).

def compute_loss_nnx:
   if num_tiles>1:
        tiling_utils.compute_tiled_loss()
   else:
       compute loss normally

Comment (NuojCheng) [reply src/MaxText/tiling_utils.py]: Make sense. I will make the change.

File: src/MaxText/tiling_utils.py (hunk 6 above)
Comment (NuojCheng) [reply src/MaxText/tiling_utils.py]: Will make the change
